commit 1a43474beae69f0b8483a1ccae866722c1334c5e
Author: John Doe <jdoe@email.com>
Date:   Sun Feb 11 23:44:53 2024 +0800

    win7dep: Remove Win7/8 code from mojo for ipcz    2023-01-20 08:31:00    3225b0b53ab0d27a0d28e2f8e5a2d3616cbded4e

diff --git a/mojo/core/embedder/embedder.cc b/mojo/core/embedder/embedder.cc
index e30f4f2af23b2..8ddfb6b072713 100644
--- a/mojo/core/embedder/embedder.cc
+++ b/mojo/core/embedder/embedder.cc
@@ -35,7 +35,12 @@
         // BUILDFLAG(IS_ANDROID)
 #endif  // !BUILDFLAG(IS_NACL)
 
-namespace mojo::core {
+#if BUILDFLAG(IS_WIN)
+#include "base/win/windows_version.h"
+#endif
+
+namespace mojo {
+namespace core {
 
 namespace {
 
@@ -74,14 +79,25 @@ void InitFeatures() {
       base::FeatureList::IsEnabled(kMojoInlineMessagePayloads));
 
   if (base::FeatureList::IsEnabled(kMojoIpcz)) {
-    EnableMojoIpcz();
+    EnableMojoIpczIfSupported();
   } else {
     g_mojo_ipcz_enabled.store(false, std::memory_order_release);
   }
 }
 
-void EnableMojoIpcz() {
-  g_mojo_ipcz_enabled.store(true, std::memory_order_release);
+void EnableMojoIpczIfSupported() {
+#if BUILDFLAG(IS_WIN)
+  // TODO(https://crbug.com/1299283): Sandboxed processes on Windows versions
+  // older than 8.1 require some extra (not yet implemented) setup for ipcz to
+  // work properly. This is omitted for early experimentation.
+  const bool kIsIpczSupported =
+      base::win::GetVersion() >= base::win::Version::WIN8_1;
+#else
+  const bool kIsIpczSupported = true;
+#endif
+  if (kIsIpczSupported) {
+    g_mojo_ipcz_enabled.store(true, std::memory_order_release);
+  }
 }
 
 void Init(const Configuration& configuration) {
@@ -168,4 +184,5 @@ IpczDriverHandle CreateIpczTransportFromEndpoint(
   return ipcz_driver::ObjectBase::ReleaseAsHandle(std::move(transport));
 }
 
-}  // namespace mojo::core
+}  // namespace core
+}  // namespace mojo
diff --git a/mojo/core/embedder/embedder.h b/mojo/core/embedder/embedder.h
index 2d31ced82a815..88645eff1b4e5 100644
--- a/mojo/core/embedder/embedder.h
+++ b/mojo/core/embedder/embedder.h
@@ -17,7 +17,8 @@
 #include "mojo/public/cpp/platform/platform_channel_endpoint.h"
 #include "third_party/ipcz/include/ipcz/ipcz.h"
 
-namespace mojo::core {
+namespace mojo {
+namespace core {
 
 // Basic configuration/initialization ------------------------------------------
 
@@ -50,13 +51,14 @@ scoped_refptr<base::SingleThreadTaskRunner> GetIOTaskRunner();
 // base::Features inside of Mojo.
 COMPONENT_EXPORT(MOJO_CORE_EMBEDDER) void InitFeatures();
 
-// Enables MojoIpcz. Called before Init() is called. Only call this if the
-// current program doesn't have base::FeatureList integration, since otherwise
-// InitFeatures() will do the work.
+// Enables MojoIpcz if it is supported on the current platform. Called before
+// Init() is called. Only call this if the current program doesn't have
+// base::FeatureList integration, since otherwise InitFeatures() will do the
+// work.
 //
 // TODO(crbug.com/1299283): Remove once MojoIpcz becomes the default
 // implementation.
-COMPONENT_EXPORT(MOJO_CORE_EMBEDDER) void EnableMojoIpcz();
+COMPONENT_EXPORT(MOJO_CORE_EMBEDDER) void EnableMojoIpczIfSupported();
 
 // Indicates whether the ipcz-based Mojo implementation is enabled. This can be
 // done by enabling the MojoIpcz feature.
@@ -88,6 +90,7 @@ IpczDriverHandle CreateIpczTransportFromEndpoint(
     const TransportEndpointTypes& endpoint_types,
     base::Process remote_process = base::Process());
 
-}  // namespace mojo::core
+}  // namespace core
+}  // namespace mojo
 
 #endif  // MOJO_CORE_EMBEDDER_EMBEDDER_H_
